### PR TITLE
Central config server transition

### LINKED ISSUE #2062: omSupply Central Server Transition
# Central Server Background

We use classic mSupply central server for 
* Configurations (store, items, master lists, etc..)
* Synchronisation (central sync hub for remote site, transfer of records between sites, backup, etc..)

When omSupply development started, we decided that it was much faster to use existing mSupply central server rather then implement new omSupply central server. This option offered other benefits, like being able to use mSupply and omSupply in the same system.

# Motivation

1. We are planning to replace mSupply central server with omSupply central server at some stage
2. New configuration UI needs to be added to mSupply for new omSupply only features (this would need to be done twice when omSupply is implemented), i.e. document/program configurations, multi pack units, asset module config.
3. New omSupply only supervisor (all store) mode features may need to be added to mSupply, i.e. global program/document reports, global asset and cold chain performance dashboard/reports ?

Although our existing strategy is to do `3` in Graphana dashboard, having these indicators directly in omSupply adds an option that may be more desirable in some use cases.

# Proposal

omSupply site can act as `central configuration` and `central data hub` for a subset of <ins>new record</ins> types. It would run in parallel with exiting mSupply remote server. omSupply remote sites would sync with both mSupply and then omSupply central server

Some research was done in [this trello card](https://trello.com/c/AJk8pLkh/26-omsupply-central-parallel) (the card and the issue will be updated with details)

<ins>new record</ins> types: It should be possible to transition to old records too, but this brings it's own challenges, for now focus of this issue is for new record types. 

# Examples

### Mutli Unit Configurations

Multi unit configuration UI can be added to omSupply central, and will sync to omSupply remote sites.

### All store mode <a href="#all-store-mode" id="all-store-mode"></a>

Any mSupply instance can access data for all stores via central omSupply API. When data for all stores needs to be accessed, it can be routed to central omSupply API rather then using current remote omSupply server. Non active store data can only be accessed when there is internet connection.

This option would require more consideration (i.e. front end consumable API of central server needs to be backwards compatible)

# Challenges

* All data needs to be present in omSupply central to make decisions about "shared" (patient related) record type routing (i.e. all name_store_joins). For new central data or new remote data this should not be a problem, since routing is pretty simple.

### PR DESCRIPTION
**Pull request by andreievg**
Fixes #2062

# 👩🏻‍💻 What does this PR do? 

* Creates a REST endpoint for central server sync API (just pull for now)
  * Authentication of site is proxied to original mSupply central server (which requires some refactoring to make SyncSettings serialisable and sharable between omSupply instances)
* Adds a check for IS_CENTRAL_SERVER env variable (to be discuss further in carry over issue) and uses it various places
* Translation for pack variant and adds pack variant to changelog
* Adds a pull from omsupply central to existing syncrhoniser
* Reshapes sync api error and push records shape to allow for existing changelog to push record to also be used by central server (with existing translators)
* Push translator now get a enum passed through to differentiate between Legacy, OmSupplyRemote/CentralSitePush (where OmSupplyCentralSitePush is for central server sync API, when central server responds to pull request) 
* Create a common cursor control (there were a few implementations of this, decided to consolidate to one common one)

I think it's worth looking at follow up [PR with doc](https://github.com/msupply-foundation/open-msupply/pull/2820) additions to understand how this work, briefly:

omSupply central server is just another remote site, but it should be hosted (or routed) on the same machine as original mSupply central, with port + 2 of original mSupply port (this is up for contention, the simples I thought for now). IS_CENTRAL_SERVER env variable is used to control if central server functionality should be on or off.

When omSupply remote site asks omSupply central for records, it passes a cursors, which is used by omSupply central to figure out what needs to be pushed from it's change log. Existing translators are used (since push translation will be the same as what is returned by pull, when the times comes to implement remote records).

Translations between omSupply central and remote are just done with serde.

# 🧪 How has/should this change been tested? 

Start omSupply with port + 2 of original mSupply port, and with IS_CENTRAL_SERVER env variable (can look at [integration test in follow up PR](https://github.com/msupply-foundation/open-msupply/pull/2819), there is a command line there). Start another mSupply instance normally. Change pack variants in omSupply central, they should sync to mSupply remote (note, for now it's possible to edit pack variants in omSupply remote, but it will be disabled)

## 💌 Any notes for the reviewer?

There are a few carry overs and consideration to discuss in a meeting, they are in [carry over issue](https://github.com/msupply-foundation/open-msupply/issues/2821), under "Questions" section.

I think if i was reviewing this i would look at sync_on_central files, then v6 files, then the changes in translations and renaming should make more sense. I can also go over the PR in person.

## 📃 Documentation

_No user facing changes_

or

- [ ] Functional change 1
- [ ] Functional change 2

### CODE REVIEW

File: server/graphql/general/src/queries/sync_status.rs
Comment (andreievg) [inline server/graphql/general/src/queries/sync_status.rs:64]: This is not added to front end yet, but in carry over issue

File: server/graphql/general/src/sync_api_error.rs
Comment (andreievg) [inline server/graphql/general/src/sync_api_error.rs:5]: Changed since it's also used in V6 now

File: server/postman/Open mSupply REST.postman_collection.json
Comment (andreievg) [inline server/postman/Open mSupply REST.postman_collection.json:1]: Check out postman collection to see the shape of request/response

File: server/repository/src/db_diesel/key_value_store.rs
Comment (andreievg) [inline server/repository/src/db_diesel/key_value_store.rs:190]: This was refactored and applied to places that use cursors, sorry this was done quite early on, I probably should have done after this PR, it adds a little bit of noise to the PR

File: server/repository/src/db_diesel/pack_variant_row.rs
Comment (andreievg) [inline server/repository/src/db_diesel/pack_variant_row.rs:37]: Pack variant is translated purely with serde (well apart from SyncTranlsation implementation for dependencies and to indicate which sites should push it

File: server/repository/src/migrations/mod.rs
Comment (andreievg) [inline server/repository/src/migrations/mod.rs:93]: Added v1_07_00 translations to pack variant, and moved all pack_variant translations there

File: server/repository/src/migrations/v1_07_00/central_omsupply.rs
Comment (andreievg) [inline server/repository/src/migrations/v1_07_00/central_omsupply.rs:27]: note, no deletion triggers

File: server/service/src/lib.rs
Comment (andreievg) [inline server/service/src/lib.rs:247]: used for cursor translation when sending in API

File: server/service/src/sync/api/common_records.rs
Comment (andreievg) [inline server/service/src/sync/api/common_records.rs:12]: I was changing SyncAction from SyncActionV5, and decided to just use rename_all, another sidetrack sorry

File: server/service/src/sync/api/core.rs
Comment (andreievg) [inline server/service/src/sync/api/core.rs:17]: Added to carry over, to make sure it's Android not mobile

File: server/service/src/sync/api/core.rs
Comment (andreievg) [inline server/service/src/sync/api/core.rs:49]: SyncApiSettings needs to be sent to omSupply central for authentication (for hardware id and credentials, thus it's shape was refactored and SyncApiCredentials was removed

File: server/service/src/sync/api/get_site_info.rs
Comment (andreievg) [inline server/service/src/sync/api/get_site_info.rs:27]: Side track, done early on

File: server/service/src/sync/api_v6/core.rs
Comment (andreievg) [inline server/service/src/sync/api_v6/core.rs:58]: Yeah, this is hacky, I agree, but alternatives was quite a bit of work ? To be discussed in the carry over issues/question

File: server/service/src/sync/api_v6/mod.rs
Comment (andreievg) [inline server/service/src/sync/api_v6/mod.rs:21]: Most types here are used in the API and are shared between omSupply central and remote, type safety should be guaranteed for the same version, but need to think about this when version changes

File: server/service/src/sync/central_data_synchroniser.rs
Comment (andreievg) [inline server/service/src/sync/central_data_synchroniser.rs:83]: I think this is a good example why I wanted to consolidate cursor controller

File: server/service/src/sync/remote_data_synchroniser.rs
Comment (andreievg) [inline server/service/src/sync/remote_data_synchroniser.rs:243]: I think it's ok to use From when it's used like RemoteSyncRecordV5::from (because can navigate here with click through, rather then the trait)

File: server/service/src/sync/sync_on_central/mod.rs
Comment (andreievg) [inline server/service/src/sync/sync_on_central/mod.rs:78]: Authenticate against mSupply central

File: server/service/src/sync/sync_on_central/mod.rs
Comment (andreievg) [inline server/service/src/sync/sync_on_central/mod.rs:84]: Added to carry over

File: server/service/src/sync/sync_on_central/mod.rs
Comment (andreievg) [inline server/service/src/sync/sync_on_central/mod.rs:101]: I like how this could be used with minimum changes

File: server/service/src/sync/central_data_synchroniser_v6.rs
Comment (jmbrunskill) [inline server/service/src/sync/central_data_synchroniser_v6.rs:49]: Is there an issue for this TODO?

File: server/service/src/sync/central_data_synchroniser_v6.rs
Comment (jmbrunskill) [inline server/service/src/sync/central_data_synchroniser_v6.rs:77]:
```suggestion
            if is_empty && total_records == 0 {
```

File: server/service/src/sync/central_data_synchroniser_v6.rs
Comment (jmbrunskill) [inline server/service/src/sync/central_data_synchroniser_v6.rs:67]: It's not important for this PR, but I feel like if catch the error instead of using `?` it should be safe to update the cursor only on  error on finishing the batch. I'm not sure if updating the cursor on every record is much overhead but feels like it could slow down the processing which I think could be a bit slow at times?
I guess it's probably more important for things like TemperatureLog Row where we might have a lot of records.

File: server/service/src/sync/sync_on_central/mod.rs
Comment (jmbrunskill) [inline server/service/src/sync/sync_on_central/mod.rs:72]: Why not just  `use SyncParsedErrorV6` personally not a big fan of this re-naming pattern. I guess it makes copy and paste easier?

File: server/service/src/sync/sync_on_central/mod.rs
Comment (jmbrunskill) [inline server/service/src/sync/sync_on_central/mod.rs:87]: I don't understand this comment?

File: server/service/src/sync/sync_on_central/mod.rs
Comment (jmbrunskill) [inline server/service/src/sync/sync_on_central/mod.rs:101]: It feels a bit weird that this function is called `Pull` but creates `Push` records.
Maybe the function could be called `pull_from_remote` that does sound a bit clunky though.
This is just me sharing my thoughts/feedback, not suggesting we need to change it...

File: server/service/src/sync/test/pull_and_push.rs
Comment (jmbrunskill) [inline server/service/src/sync/test/pull_and_push.rs:73]:
```suggestion
        get_all_omsupply_central_push_records(),
```

File: server/service/src/sync/translations/mod.rs
Comment (jmbrunskill) [inline server/service/src/sync/translations/mod.rs:250]: Should comment be after the  `OmSupplyCentralSitePush` a bit confused if this all relates to    `OmSupplyRemoteSitePush`?

File: server/service/src/sync/translations/mod.rs
Comment (jmbrunskill) [inline server/service/src/sync/translations/mod.rs:289]: Is `unimplemented!()` a good idea here? Feels a bit risky, but I can see it's maybe better then returning true or false...

**Review by jmbrunskill — APPROVED** [review]
Looks good to me, but haven't tested myself.
Figure we can refine/test a bit more as we add new features for CCE for example.

I would be keen to see or perhaps do a test branch where we try out a feature flag in rust rather than `is_central` server, doesn't feel like the `is_central` thing is creating spaghetti code so far.

I don't have my head deep enough into the space, but feels like where remote servers could send records for sites they don't own or something checks for that kind of thing might get a bit messier, but I'm personally happy to go with what you have here. Doesn't feel like it would be too hard to migrate to using a rust feature flag if we wanted too.

File: server/service/src/sync/api_v6/core.rs
Comment (jmbrunskill) [inline server/service/src/sync/api_v6/core.rs:9]: Still interested in another solution rather than port offset. Doesn't look too hard to plumb another url through IMHO.
Even if we default to the port offset, but allow it to be set in a configuration that would be a good thing.
I think having a separate port as an option will make dev's life much easier. We don't all want to have to setup Caddy or something...

If rust could be the proxy for old mSupply that would be a good solution. I'm keen for a follow up issue, and happy to experiment with options if you'd like.

Comment (clemens-msupply) [reply server/service/src/sync/api_v6/core.rs:9]: If I understood correctly, we have the use-case that we need to run the remote server and the new central server on the same machine, and then we need two ports, don't we?

Think as long as we can configure the url/port we can decide later how we deploy it and what is best for a dev setup.

File: server/service/src/sync/translations/mod.rs
Comment (roxy-dao) [reply server/service/src/sync/translations/mod.rs:250]: Also central (:

File: server/service/src/sync/central_data_synchroniser_v6.rs
Comment (andreievg) [reply server/service/src/sync/central_data_synchroniser_v6.rs:49]: Nope, this comment has been there already, i just added to [original sync carry over](https://github.com/msupply-foundation/open-msupply/issues/608), thanks

File: server/service/src/sync/central_data_synchroniser_v6.rs
Comment (andreievg) [reply server/service/src/sync/central_data_synchroniser_v6.rs]: Thanks, cc08fbe03ce8dbc6b91a8bbcf2e97931ef784075

File: server/service/src/sync/central_data_synchroniser_v6.rs
Comment (andreievg) [reply server/service/src/sync/central_data_synchroniser_v6.rs:67]: The idea here was we update sync buffer and cursor atomically (if one fails the other fails). I added to this issue: https://github.com/msupply-foundation/open-msupply-internal/issues/39#issuecomment-1940314362

File: server/service/src/sync/sync_on_central/mod.rs
Comment (andreievg) [reply server/service/src/sync/sync_on_central/mod.rs:72]: > I guess it makes copy and paste easier

That and less verbose down the line, especially when used in chain mapping with `map_err` (most likely the latter)

File: server/service/src/sync/sync_on_central/mod.rs
Comment (andreievg) [reply server/service/src/sync/sync_on_central/mod.rs]: Neither do i hehe, i did a self review but missed some things, thanks: dcbd771360dd4f616bbebc49acbd30c5d80634e9

File: server/service/src/sync/sync_on_central/mod.rs
Comment (andreievg) [reply server/service/src/sync/sync_on_central/mod.rs:100]: I was hoping the OmSupplyCentralSitePush type would make sense (plus the comments for that variant). 

I also found it confusing.

How about I rename translators and associated methods to: "from_sync_record", "to_sync_record" ?

Comment (andreievg) [reply server/service/src/sync/sync_on_central/mod.rs:100]: And of course add some comment to the "pull" method here to say it's used by central API in response to a pull records request from remote site

File: server/service/src/sync/api_v6/core.rs
Comment (andreievg) [reply server/service/src/sync/api_v6/core.rs:9]: I think the main reason i went with port offset for now, was because of existing sites. How do we point them to the omSupply central server instance unless we caddy or port offset ? 

> I'm keen for a follow up issue, and happy to experiment with options if you'd like.

Sure although I am wondering for this feature to go through, what is the minimum that is needed now, and what is the strategy for existing sites for later ?

Use case, existing deployment, and we upgrade them to omSupply version that requires omSupply central (not the pack variant version but say the assets version). We don't do the port offset thing, or the caddy thing, how do we tell all remotes sites that server is now somewhere else ? 

(I think we would need classic mSupply central to expose API to query for omSupply central server location, and this would happen as the first step of sync)

File: server/repository/src/db_diesel/key_value_store.rs
Comment (andreievg) [reply server/repository/src/db_diesel/key_value_store.rs:195]: Hmm, i think the first is both shorter and clearer, won't change for now, but will take note

File: server/service/src/sync/api_v6/mod.rs
Comment (andreievg) [reply server/service/src/sync/api_v6/mod.rs:78]: > There is a spelling error in response

a64540d2926549d8c4586fa380fced819b0fe2c1, thanks

> The variant naming is inconsistent "Parsed" vs "Parsing"...

I thought they are different, since 'parsing' error is before it's parsed, and parsed error indicates we matched an error from api response. Not 100% happy with those variants, but not sure what's the best alternative

File: server/server/src/sync_on_central/mod.rs
Comment (andreievg) [reply server/server/src/sync_on_central/mod.rs:15]: I was wondering how to differentiate between things that are done by central server and remote server sync to central, that's why i did sync_on_central (including folder names in various places).

File: server/server/src/sync_on_central/mod.rs
Comment (andreievg) [reply server/server/src/sync_on_central/mod.rs]: hmm, good point, I think there would be other rest API that are central specific, sounds like we should. thanks
2729e1dea977dc80f2b7a8f055cf3f378ca764f5

File: server/service/src/sync/api_v6/core.rs
Comment (andreievg) [reply server/service/src/sync/api_v6/core.rs]: 73d8b6965f4d6a391c59e330d0efd49ee529fe02

File: server/service/src/sync/central_data_synchroniser_v6.rs
Comment (andreievg) [reply server/service/src/sync/central_data_synchroniser_v6.rs:66]: I am not sure about this comment, removed: 46a019ce175d7b4ede4d0c4bb1251c25de475dea

File: server/service/src/sync/mod.rs
Comment (andreievg) [reply server/service/src/sync/mod.rs]: Thanks, yeah better place in API, ideally in some shared place like api/mod, api/v5/, api/v6, that's a bit too much to refactor right now, I think it would also mean we can move some of the errors and sync record shapes to a common place, add to [carry over](https://github.com/msupply-foundation/open-msupply/issues/2821)

`Relocate common API shapes to a shared space`

File: server/service/src/sync/test/pull_and_push.rs
Comment (andreievg) [reply server/service/src/sync/test/pull_and_push.rs]: Nope, removed: fd5ea47e32a411a24857234338b9dd7513facdad

File: server/util/src/lib.rs
Comment (andreievg) [reply server/util/src/lib.rs:33]: It was a bit of a mission passing it around, and I wasn't yet sure how this would be specified (via feature flag, env variable, a settings that syncs etc..). Also was thinking about global multi read one writer mutex for this sort of thing, added to [carry over](https://github.com/msupply-foundation/open-msupply/issues/2821) as comment

File: server/util/src/lib.rs
Comment (andreievg) [reply server/util/src/lib.rs]: 7c2b13d8022c280f4ca4118f8fea4807636724cf, thanks

File: server/repository/src/db_diesel/key_value_store.rs
Comment (andreievg) [reply server/repository/src/db_diesel/key_value_store.rs]: Agree done here: d92564a6175eefdd581de199bcfd2659738a8ca4

File: server/service/src/sync/api_v6/mod.rs
Comment (andreievg) [reply server/service/src/sync/api_v6/mod.rs:21]: Yep, I think keeping separate API version for backward compatibility of all sites would have way too many use cases to keep track of, I think we would take similar approach to what is being done for legacy mSupply central sync api, try to keep it backwards compatible, if we do break compatibility it would require all sites to update to new version. I think this would be possible if we 'sync' binary to remote sites, something that has been mentioned, maybe for RnD day

**Comment by andreievg** [comment]
Thank you for your reviews

@jmbrunskill 

> I would be keen to see or perhaps do a test branch where we try out a feature flag in rust rather than is_central server, doesn't feel like the is_central thing is creating spaghetti code so far.

Hmm, as per discussion in our meeting, the advantage of feature flag would be to build seperate binary, which is usefull when app is being installed, we do have installers that can add env variable to service startup. I am still trying to figure out exactly why feature flag woudl be useful

@clemens-msupply 

> Looks good! Mostly some spelling fixes. Think its better to only read the env variable once at startup and put it into the settings service or similar....

There is a comment about this above and added to carry over

@jmbrunskill @clemens-msupply 

There were a few questions and confusions about push translations used for central and remote, I did a branch where I renamed things to try and make it easier to understand, it's this one: https://github.com/msupply-foundation/open-msupply/compare/Central-config-server-transition...Central-config-server-transition(Renaming-translations), I didn't want to add confusion to PR review by adding loads more changes, so though to keep it separate for now, if you agree will merge into this branch before merging this branch

You can comment in the commit: https://github.com/msupply-foundation/open-msupply/commit/44b072420323584b423870da3e7c7b926f7d5703

The main change is:

https://github.com/msupply-foundation/open-msupply/blob/44b072420323584b423870da3e7c7b926f7d5703/server/service/src/sync/translations/mod.rs#L243-L257

and here:

https://github.com/msupply-foundation/open-msupply/blob/44b072420323584b423870da3e7c7b926f7d5703/server/service/src/sync/translations/mod.rs#L259-L333

I still like the idea of using one translation trait for all translation operation on a particular entity (remote pull and push to legacy, omSupply central, and when omSupply central is translating records in response to pull request from remote site).

How does that one look ?

I think apart form that I dealt with all of the PR comments, thanks once again for review and re-review

**Comment by jmbrunskill** [comment]
> How does that one look ?

I like this change, removes confusion about pushing/pulling

File: server/util/src/lib.rs
Comment (clemens-msupply) [reply server/util/src/lib.rs:33]: Was mainly thinking about alterntive setups, e.g when creating a test setup where you might not want to use an env variable. For things like that it might be useful have configurable.

Having a global variable for it is probably ok but still have the feeling its "cleaner" to pass it arround, e.g. the immutable settings object we already have seems suitable and is already available through the SettingService.

**Review by clemens-msupply — APPROVED** [review]
Thanks for the changes!

**Comment by andreievg** [comment]
Thanks for your reviews, much appreciated, there is still some things to discuss, but they should all be captured in the carry over

**Pull request merged**
